pandas/io/sql.py

```diff
@@ -840,6 +867,181 @@ def create(self):
         else:
             self._execute_create()
 
+    def _load_existing_pkeys(self, primary_keys, primary_key_values):
+        """
+        Load existing primary keys from Database
+
+        Parameters
+        ----------
+        primary_keys : list of str
+            List of primary key column names
+        primary_key_values : list of str
+            List of primary key values already present in incoming dataframe
+
+        Returns
+        -------
+        list of str
+            primary key values in incoming dataframe which already exist in database
+        """
+        from sqlalchemy import (
+            and_,
+            select,
+        )
+
+        cols_to_fetch = [self.table.c[key] for key in primary_keys]
+        # select_stmt = select(cols_to_fetch).where(
+        #     tuple_(*cols_to_fetch).in_(primary_key_values)
+        # )
+        select_stmt = select(cols_to_fetch).where(
+            and_(
+                col.in_(key[i] for key in primary_key_values)
+                for i, col in enumerate(cols_to_fetch)
+            )
+        )
+        return self.pd_sql.execute(select_stmt).fetchall()
+
+    def _split_incoming_data(self, primary_keys, keys_in_db):
+        """
+        Split incoming dataframe based off whether primary key already exists in db.
+
+        Parameters
+        ----------
+        primary_keys : list of str
+            Primary keys columns
+        keys_in_db : list of str
+            Primary key values which already exist in database table
+
+        Returns
+        -------
+        tuple of DataFrame, DataFrame
+            DataFrame of rows with duplicate pkey, DataFrame of rows with new pkey
+        """
+        in_db = _wrap_result(data=keys_in_db, columns=primary_keys)
+        # Get temporary dataframe so as not to delete values from main df
+        temp = self._get_index_formatted_dataframe()
+        exists_mask = (
+            temp[primary_keys]
+            .apply(tuple, axis=1)
+            .isin(in_db[primary_keys].apply(tuple, axis=1))
+        )
```

Comment:
Can we do a `DataFrame.merge` here then query based on `_merge` using the `indicator=True` argument? I am afraid using `.apply(tuple, axis=1)` twice can get pretty slow for large dataframes.

Reply:
Or some other options for `isin` on multiple columns to keep it vectorized: https://stackoverflow.com/questions/45198786/how-to-use-pandas-isin-for-multiple-columns

I can make a suggestion if you want.

Reply:
Thanks @erfannariman , that's a great point.  I have decided to try the multiindex `isin` approach, as it seems to scale better than the merge (which, if I remember correctly, can also be a bit more memory hungry).